# Assignment 2
**Author:** F.F. Höhne Tarragona _S4605632_ <br>
**Course:** Smart Industry Operations _EBM223A05.2021-2022.1B_ <br>
**Lecturer:** Dr Christos Emmanouilidis <br>

**2021-12-17**


For this assignment, we are requested to pre-select potential employees for a manufactuing organisaton. 
For the selected potential employees; a list including their ocmplete record is to be returned in uniform format. 

The following code presents a method to select employees for the **Finance Department**. 

The following steps are to be considered:
1. Add a column for the intended department of the  employee: Finance 
2. Prospective employees should be from the European Union, Norway, and the  UK. 
3. Specifically, applicants must meet the following requirements:
    1. have skills at least in Planning 
    2. a minimum of 4 years  experience
    3. the education level should be above a BSc. degree (as specified in the Assignment Outline Doc.)
    4. employees should not be at the lowest two tiers of performance (D & E) 
    5. spoken language should be three out of Dutch, English, French, Spanish

-----

## Step 1: Reading the data
As a first step, we import the needed librabries and we open the dataset. 

In [1]:
import pandas as pd
import numpy as np
from countrygroups import EUROPEAN_UNION

In [2]:
# read in data specifying seperator/delimiter
employeeData = pd.read_csv('DataAssignment2.csv', sep=',')

We then do a preliminary analysis of the data, looking at the shape, type, and form of data; and getting an idea of the values for each attribute. 

It is important to look for consistency and lack there of, as well as null values in the table. 

In [3]:
employeeData.shape

(1000, 21)

In [4]:
employeeData.head(10)

RecID First Name      Surname  Gender  Age Date of birth  \
0  0660900YMA       Kurt       Jensen    male   62    04/07/1959   
1  5612835CHK      Joann     Woodward  female   39    11/11/1982   
2  6313047ALD     Thomas      Belcher  female   44    25/12/1976   
3  4216935AQI   Patricia  Cirrincione  female   39    22/08/1982   
4  2765119WTR   Reginald         Fogg    male   54    11/09/1967   
5  5863686WTY  Christina        Butts  female   62    11/08/1959   
6  5170070JEI     Freida       Bailey  female   50    30/11/1971   
7  8127770BEB     Gloria       Twyman  female   30    18/02/1991   
8  9980683KSC     Robert   Turnipseed    male   64    04/12/1957   
9  9830617VWN      Sarah       Torres  female   41    03/02/1980   

   Years of experience  Education Level                   Country Performance  \
0                   10           master  Republic of South Africa          94   
1                    9  upper secondary                     Italy           C   
2                   17        technical                       USA          89   
3                   13           master                    Brazil           C   
4                   21        doctorate                    Sweden           B   
5                   23         bachelor                     Malta          24   
6                   24        technical                    France          11   
7                    0        technical                   Romania          55   
8                   17           master                   Belgium          50   
9                   15        technical                     Malta           E   

   ... Machining Welding Additive Painting Control Quality Planning Services  \
0  ...       YES     YES      YES       NO      NO      NO      YES       NO   
1  ...       YES      NO       NO       NO      NO      NO      YES       NO   
2  ...        NO     YES      YES      YES     YES      NO      YES      YES   
3  ...       YES     YES      YES       NO      NO     YES      YES      YES   
4  ...        NO     YES       NO       NO      NO     YES       NO      YES   
5  ...        NO     YES      YES       NO      NO     YES      YES       NO   
6  ...       YES     YES      YES      YES      NO      NO      YES       NO   
7  ...        NO     YES      YES       NO      NO      NO       NO       NO   
8  ...        NO     YES       NO       NO     YES      NO       NO      YES   
9  ...        NO      NO       NO      YES      NO     YES       NO      YES   

  Shipping spoken languages  
0       NO  EN,SP,NL,GE,IT,  
1      YES      EN,F,NL,GE,  
2       NO        NL,GE,IT,  
3      YES   EN,F,NL,GE,IT,  
4       NO            F,SP,  
5      YES           NL,IT,  
6      YES      F,NL,GE,IT,  
7       NO           NL,GE,  
8       NO               F,  
9      YES           EN,SP,  

[10 rows x 21 columns]

In [5]:
# dtype should be in accordance with data input; e.g. Age - int64, Date of birth - date
employeeData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   RecID                1000 non-null   object
 1   First Name           1000 non-null   object
 2   Surname              1000 non-null   object
 3   Gender               1000 non-null   object
 4   Age                  1000 non-null   int64 
 5   Date of birth        1000 non-null   object
 6   Years of experience  1000 non-null   int64 
 7   Education Level      1000 non-null   object
 8   Country              1000 non-null   object
 9   Performance          1000 non-null   object
 10  Assembly             1000 non-null   object
 11  Machining            1000 non-null   object
 12  Welding              1000 non-null   object
 13  Additive             1000 non-null   object
 14  Painting             1000 non-null   object
 15  Control              1000 non-null   object
 16  Quality

In [6]:
# check RedID is unique for all people
len(employeeData['RecID'].unique())

1000

In [7]:
# Gender should be named equally
employeeData['Gender'].unique()

array(['male', 'female', 'f', 'm'], dtype=object)

In [8]:
# Age should be between 18 and 67 (retirement age)
employeeData['Age'].describe()

count    1000.000000
mean       43.254000
std        14.031121
min        18.000000
25%        31.000000
50%        43.000000
75%        55.000000
max        67.000000
Name: Age, dtype: float64

In [9]:
# check no negative values, no values look out of the norm 
employeeData['Years of experience'].describe()

count    1000.000000
mean       10.630000
std         9.605695
min        -1.000000
25%         2.000000
50%         9.000000
75%        18.000000
max        35.000000
Name: Years of experience, dtype: float64

In [10]:
# check different levels of education level and consistency in naming
employeeData['Education Level'].describe() 
employeeData['Education Level'].unique() 

array(['master', 'upper secondary', 'technical', 'doctorate', 'bachelor'],
      dtype=object)

In [11]:
# check different countries of origin and consistency in naming
employeeData['Country'].describe() 
employeeData['Country'].unique() 

array(['Republic of South Africa', 'Italy', 'USA', 'Brazil', 'Sweden',
       'Malta', 'France', 'Romania', 'Belgium', 'Slovakia', 'Luxembourg',
       'Ireland', 'Republic of Cyprus', 'Finland', 'Austria', 'Poland',
       'Latvia', 'Hungary', 'Estonia', 'Slovenia', 'Czech Republic',
       'China', 'Germany', 'Denmark', 'Bulgaria', 'Portugal', 'Lithuania',
       'Greece', 'Australia', 'Spain', 'Netherlands', 'Republic of Congo',
       'United Kingdom', 'Croatia'], dtype=object)

In [12]:
# should be between A-F; however we have 105 unique values. Perfromance uses different scales (half use an alphabetic scale) 
employeeData['Performance'].describe()
employeeData['Performance'].unique()
employeeData['Performance'].value_counts()

A     110
E     108
D      99
C      98
B      81
     ... 
18      2
75      2
96      2
35      1
51      1
Name: Performance, Length: 105, dtype: int64

In [13]:
# naming should be consistent
# needs further work for a propoer analysis as differnt languages are listed and are therefore difficult to analyze
employeeData['spoken languages'].describe()

count             985
unique             63
top       F,SP,NL,IT,
freq               23
Name: spoken languages, dtype: object

------

## Step 2: Data cleaning 
From the previous steps; we see that: 
  1. some candidates have not added their spoken languages - NaNs
  2. overall, the naming of the attirbutes is clear and free of homonyms or synonyms. However:
      1. the label for the attibute "spoken languages" is inconsisten with the rest (lower case)
      2. the label for "Education Level" also doesn't match
      3. other changes in the table formatting needed:
         - Changing "Date of birth" Dtype to DateTime
         - Removing negative values from "Years of experience"
         - Adding the needed column for "Intended department"

  3. Regarding conversion needs: 
      1. "Gender" needs to be converted for uniform naming 
      2. EU countries need to be grouped for further filtering 
      3. "Performance levels" need to be converted into one scale




#### 1. Remove NaNs 

In [14]:
## 1. as there is a min. requirement for this attribute, all NaNs can be removed
df = employeeData.dropna()

#### Change table formatting/ naming

In [15]:
# correcting column names
df = df.rename(columns={'Education Level':'Education level'})
df = df.rename(columns={'spoken languages':'Spoken languages'})

#  correcting DType
df['Date of birth'] = pd.to_datetime(df['Date of birth'])
# removing negative values from Years of experience
df['Years of experience']=df['Years of experience'].clip(lower=0)
# adding "Intended department", changing Dtype
df['Intended department'] = np.nan

In [16]:
# check changes are correct
# check for correct input/ no unnusual answers
df.info()
df['Date of birth'].describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 985 entries, 0 to 999
Data columns (total 22 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   RecID                985 non-null    object        
 1   First Name           985 non-null    object        
 2   Surname              985 non-null    object        
 3   Gender               985 non-null    object        
 4   Age                  985 non-null    int64         
 5   Date of birth        985 non-null    datetime64[ns]
 6   Years of experience  985 non-null    int64         
 7   Education level      985 non-null    object        
 8   Country              985 non-null    object        
 9   Performance          985 non-null    object        
 10  Assembly             985 non-null    object        
 11  Machining            985 non-null    object        
 12  Welding              985 non-null    object        
 13  Additive             985 non-null  

C:\Users\fabia\AppData\Local\Temp/ipykernel_7792/2244238694.py:4: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df['Date of birth'].describe()


count                     985
unique                    964
top       1957-04-21 00:00:00
freq                        2
first     1954-02-28 00:00:00
last      2003-12-10 00:00:00
Name: Date of birth, dtype: object

#### 3. Converting values to uniform scales

In [17]:
# uniform gender naming 
df = df.replace({'Gender': {'f': 'female', 'm': 'male'}})

In [18]:
# check changes are correct
df.Gender.unique()

array(['male', 'female'], dtype=object)

In [19]:
# EU countries need to be grouped for further filtering 
lst = EUROPEAN_UNION.names
c = lambda x: 'EU' if x in lst else x
df['Country'] = df['Country'].map(c)

In [20]:
# check changes are correct
df.head()

RecID First Name      Surname  Gender  Age Date of birth  \
0  0660900YMA       Kurt       Jensen    male   62    1959-04-07   
1  5612835CHK      Joann     Woodward  female   39    1982-11-11   
2  6313047ALD     Thomas      Belcher  female   44    1976-12-25   
3  4216935AQI   Patricia  Cirrincione  female   39    1982-08-22   
4  2765119WTR   Reginald         Fogg    male   54    1967-11-09   

   Years of experience  Education level                   Country Performance  \
0                   10           master  Republic of South Africa          94   
1                    9  upper secondary                        EU           C   
2                   17        technical                       USA          89   
3                   13           master                    Brazil           C   
4                   21        doctorate                        EU           B   

   ... Welding Additive Painting Control Quality Planning Services Shipping  \
0  ...     YES      YES       NO      NO      NO      YES       NO       NO   
1  ...      NO       NO       NO      NO      NO      YES       NO      YES   
2  ...     YES      YES      YES     YES      NO      YES      YES       NO   
3  ...     YES      YES       NO      NO     YES      YES      YES      YES   
4  ...     YES       NO       NO      NO     YES       NO      YES       NO   

  Spoken languages Intended department  
0  EN,SP,NL,GE,IT,                 NaN  
1      EN,F,NL,GE,                 NaN  
2        NL,GE,IT,                 NaN  
3   EN,F,NL,GE,IT,                 NaN  
4            F,SP,                 NaN  

[5 rows x 22 columns]

In [21]:
# as the grades following the alphabetical scale offer less information than the numerical scaling, 
# the alphabetical scale will be kept to avoid creating missleading data points 
df = df.replace({'Performance' : { 'A' : 90, 'B' : 70, 'C' : 50 , 'D' : 30 , 'E' : 10}})

In [22]:
# check for correct data
df['Performance'].describe()

count     985
unique    105
top        90
freq      109
Name: Performance, dtype: int64

In [23]:
# group accordingly to alphabetical scale 
df['Performance']=pd.to_numeric(df['Performance'])
Grade = pd.cut(df.Performance, bins=[0,20,40,60,80,100], labels=['E','D','C','B','A'])

In [24]:
# add to dataframe
df.insert(9,'Grade', Grade)

In [25]:
# check for correcness
df['Grade'].unique()

['A', 'C', 'B', 'D', 'E', NaN]
Categories (5, object): ['E' < 'D' < 'C' < 'B' < 'A']

----

## Step 3: Data analysis 
As described in the introduction, the data needs to be filtered to find the possible candidates. 
The conditions for this step can be found in the introduction section.

In [26]:
# min. education requirement: BSc. 
# prospective employees should be from the European Union, Norway, and the  UK.
# skills in at least planning 
# 4 or more years experience
lst_countries = ['EU', 'United Kingdom', "Norway"]
lst_edu = ['bachelor','master','doctorate']

df_filtered_1 = df.loc[((df['Education level'].isin(lst_edu)) & df['Country'].isin(lst_countries)) & (df['Planning']== "YES") & (df['Years of experience'] >= 4)]

In [27]:
# check for correct filtering country
df_filtered_1['Country'].unique()
# check for correct filtering planning
df_filtered_1['Planning'].unique()
# check for correct filtering experience -- look at min.  
df_filtered_1['Years of experience'].describe()

count    164.000000
mean      14.658537
std        8.556875
min        4.000000
25%        7.000000
50%       13.000000
75%       21.000000
max       34.000000
Name: Years of experience, dtype: float64

In [28]:
# Language - min 3 - out of NL, EN, F, SP
# for min. 3 languages to be in the field, the min. string length needs to be 8 (3 commas, no spaces, 5 letters)
# then use if function to filter out if the needed languages are present
min_len = 8
lst_lan = ['NL', 'EN', 'F', 'SP']

df_filtered_2 = df_filtered_1[df_filtered_1['Spoken languages'].str.len() >= min_len]

In [29]:
# as any combination of languages is acceptable, a count has been added to ensure at least 3 of the languages are spoken. 
for x in lst_lan:
    df_filtered_2[x] = df_filtered_2['Spoken languages'].apply(lambda i: 1 if x in i else 0)

df_filtered_3 = df_filtered_2.loc[(df_filtered_2[lst_lan].sum(1, numeric_only=True) >= 3)]

C:\Users\fabia\AppData\Local\Temp/ipykernel_7792/840658950.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered_2[x] = df_filtered_2['Spoken languages'].apply(lambda i: 1 if x in i else 0)


The last of the requirements for employees is that they should not be at the lowest two tiers of performance (i.e. they need to be in the top 1/3 of applicants). 
As we only 1/2 of applcants have sent in numerical grades, alphabetical grading will be used to filter out the top third. 

The conversion bases on the assumption that alphabetical grades are the equivalent to the following scheme: 
- A: 100-80%
- B: 79-60% 
- C: 59-40% 
- D: 39-20% 
- E: >19% 

It is therefore assumed, that the top 1/3 of applicants are those with a B or more. <br> _(The numerical threshold for the given dataset is at a performance of 75%; which falls within the B bucket)._

In [30]:
lst_threshold = ['A', 'B', 'C']
df_filtered_4 = df_filtered_3[df_filtered_2['Grade'].isin(lst_threshold)]

C:\Users\fabia\AppData\Local\Temp/ipykernel_7792/3248545005.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_filtered_4 = df_filtered_3[df_filtered_2['Grade'].isin(lst_threshold)]


----
## Step 4: restore formatting and save file
Finally, we do a last check of the applicants records.

We then remove any added columns, and ensure the formatting is in line with the initial formatting. 

The last step is then to save the file as a .csv. 

In [31]:
df_filtered_4.RecID.count()

28

In [32]:
df_filtered_4 = df_filtered_4.drop(columns = lst_lan)

In [33]:
df_filtered_4.to_csv('DataAssignment2_cleaned.csv', index=False)

----
### Conclusion:

<div class="alert alert-block alert-success">
<b>Final results:</b> <br>A total of 28 applicants fill all criteria presented by the company:<br>
        - Average years of experience: 16 <br>
        - Average age: 50 <br>
        - 64% are female <br>
        - 28% have a C performance <br>
        - 25% have a B perfromance <br>
        - 46% have a A perfromance <br>
        - All are EU citizens <br>
        - Most hold a doctorate degree (43%) <br>
</div>


From this step onwards, further steps to find the best candidate include:
- Sort by total number of skills | sort by another specific skill (Services, Shipping, etc.)
- Further sort or filter by Education level
- Review max. age; to ensure lower employee turnover

Additionallly, if specific skills are not necessary for the job (e.g. welding), it would be recommended to remove given columns from database to improve readability, and reduce size of dataset. 

_At this point, no further chages will be made to the dataset, and no further data will be removed without the consensus from the HR department, to ensure no valuable data is being removed._ 